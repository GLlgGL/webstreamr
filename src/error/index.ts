import winston from 'winston';
import { BlockedReason, Context } from '../types';
import { BlockedError } from './BlockedError';
import { HttpError } from './HttpError';
import { QueueIsFullError } from './QueueIsFullError';
import { TimeoutError } from './TimeoutError';
import { TooManyRequestsError } from './TooManyRequestsError';
import { TooManyTimeoutsError } from './TooManyTimeoutsError';

export * from './BlockedError';
export * from './HttpError';
export * from './NotFoundError';
export * from './QueueIsFullError';
export * from './TimeoutError';
export * from './TooManyRequestsError';
export * from './TooManyTimeoutsError';

export const logErrorAndReturnNiceString = (ctx: Context, logger: winston.Logger, source: string, error: unknown): string => {
  if (error instanceof BlockedError) {
    if (error.reason === BlockedReason.cloudflare_challenge) {
      logger.warn(`${source}: Request was blocked via Cloudflare challenge.`, ctx);
    } else if (error.reason === BlockedReason.cloudflare_censor) {
      logger.warn(`${source}: Request was censored by Cloudflare.`, ctx);
    } else if (error.reason === BlockedReason.media_flow_proxy_auth) {
      return '⚠️ MediaFlow Proxy authentication failed. Please set the correct password.';
    } else {
      logger.warn(`${source}: Request was blocked, headers: ${JSON.stringify(error.headers)}.`, ctx);
    }

    return `⚠️ Request was blocked. Reason: ${error.reason}`;
  }

  if (error instanceof TooManyRequestsError) {
    logger.warn(`${source}: Rate limited for ${error.retryAfter} seconds.`, ctx);

    return '🚦 Request was rate-limited. Please try again later or consider self-hosting.';
  }

  if (error instanceof TooManyTimeoutsError) {
    logger.warn(`${source}: Too many timeouts.`, ctx);

    return '🚦 Too many recent timeouts. Please try again later.';
  }

  if (
    error instanceof TimeoutError
    || (error instanceof DOMException && error.name === 'TimeoutError') // sometimes this gets through, no idea why..
  ) {
    logger.warn(`${source}: Request timed out.`, ctx);

    return '🐢 Request timed out.';
  }

  if (error instanceof QueueIsFullError) {
    logger.warn(`${source}: Request queue is full.`, ctx);

    return '⏳ Request queue is full. Please try again later or consider self-hosting.';
  }

  if (error instanceof HttpError) {
    logger.error(`${source}: HTTP status ${error.status} (${error.statusText}), headers: ${JSON.stringify(error.headers)}, stack: ${error.stack}.`, ctx);

    if (error.status >= 500) {
      return `❌ Remote server has issues. We can't fix this, please try later again.`;
    }

    return `❌ Request failed with status ${error.status} (${error.statusText}). Request-id: ${ctx.id}.`;
  }

  const cause = (error as Error & { cause?: unknown }).cause;
  logger.error(`${source} error: ${error}, cause: ${cause}, stack: ${(error as Error).stack}`, ctx);

  return `❌ Request failed. Request-id: ${ctx.id}.`;
};
