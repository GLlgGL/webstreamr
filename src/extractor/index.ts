import { envGet, Fetcher } from '../utils';
import { DoodStream } from './DoodStream';
import { Dropload } from './Dropload';
import { ExternalUrl } from './ExternalUrl';
import { Extractor } from './Extractor';
import { Fastream } from './Fastream';
import { FileLions } from './FileLions';
import { FileMoon } from './FileMoon';
import { HubCloud } from './HubCloud';
import { KinoGer } from './KinoGer';
import { LuluStream } from './LuluStream';
import { Mixdrop } from './Mixdrop';
import { SaveFiles } from './SaveFiles';
import { StreamEmbed } from './StreamEmbed';
import { Streamtape } from './Streamtape';
import { SuperVideo } from './SuperVideo';
import { Uqload } from './Uqload';
import { VidSrc } from './VidSrc';
import { VixSrc } from './VixSrc';
import { Voe } from './Voe';
import { YouTube } from './YouTube';

export * from './Extractor';
export * from './ExtractorRegistry';

export const createExtractors = (fetcher: Fetcher): Extractor[] => {
  const disabledExtractors = envGet('DISABLED_EXTRACTORS')?.split(',') ?? [];

  return [
    new DoodStream(fetcher),
    new Dropload(fetcher),
    new Fastream(fetcher),
    new FileLions(fetcher),
    new FileMoon(fetcher),
    new HubCloud(fetcher),
    new KinoGer(fetcher),
    new LuluStream(fetcher),
    new Mixdrop(fetcher),
    new SaveFiles(fetcher),
    new StreamEmbed(fetcher),
    new Streamtape(fetcher),
    new SuperVideo(fetcher),
    new Uqload(fetcher),
    new VidSrc(fetcher, [ // https://vidsrc.domains/
      'bz',
      'do',
      'gd',
      'in',
      'io',
      // 'me', // redirects to .net
      'mn',
      'net',
      'tw',
      // 'vc', // 500 internal server error
      'xyz',
    ]),
    new VixSrc(fetcher),
    new Voe(fetcher),
    new YouTube(fetcher),
    new ExternalUrl(fetcher), // fallback extractor which must come last
  ].filter(extractor => !disabledExtractors.includes(extractor.id));
};
