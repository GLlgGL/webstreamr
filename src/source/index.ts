import { Fetcher } from '../utils';
import { CineHDPlus } from './CineHDPlus';
import { Cuevana } from './Cuevana';
import { Einschalten } from './Einschalten';
import { Eurostreaming } from './Eurostreaming';
import { FourKHDHub } from './FourKHDHub';
import { Frembed } from './Frembed';
import { FrenchCloud } from './FrenchCloud';
import { HomeCine } from './HomeCine';
import { KinoGer } from './KinoGer';
import { MegaKino } from './MegaKino';
import { MeineCloud } from './MeineCloud';
import { MostraGuarda } from './MostraGuarda';
import { Movix } from './Movix';
import { PrimeWire } from './PrimeWire';
import { Source } from './Source';
import { StreamKiste } from './StreamKiste';
import { VerHdLink } from './VerHdLink';
import { VidSrc } from './VidSrc';
import { VixSrc } from './VixSrc';
import { XPrime } from './XPrime';

export * from './Source';

export const createSources = (fetcher: Fetcher): Source[] => [
  // multi
  new FourKHDHub(fetcher),
  new VixSrc(fetcher),
  // EN
  new PrimeWire(fetcher),
  // new Soaper(fetcher), // "temporarily" rate-limited for over a week
  new VidSrc(fetcher),
  new XPrime(fetcher),
  // ES / MX
  new CineHDPlus(fetcher),
  new Cuevana(fetcher),
  new HomeCine(fetcher),
  new VerHdLink(fetcher),
  // DE
  new Einschalten(fetcher),
  new KinoGer(fetcher),
  new MegaKino(fetcher),
  new MeineCloud(fetcher),
  new StreamKiste(fetcher),
  // FR
  new Frembed(fetcher),
  new FrenchCloud(fetcher),
  new Movix(fetcher),
  // IT
  new Eurostreaming(fetcher),
  new MostraGuarda(fetcher),
];
