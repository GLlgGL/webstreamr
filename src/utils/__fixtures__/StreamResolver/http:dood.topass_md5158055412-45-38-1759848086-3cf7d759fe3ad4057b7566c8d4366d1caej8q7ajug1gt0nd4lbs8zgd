https://do365ki.cloudatacdn.com/u5kjzcovmpd3sdgge7tc4jacdzaikwer6ff4qd463b6qj6lodlgpnsbkffvq/xy59ygfkku~