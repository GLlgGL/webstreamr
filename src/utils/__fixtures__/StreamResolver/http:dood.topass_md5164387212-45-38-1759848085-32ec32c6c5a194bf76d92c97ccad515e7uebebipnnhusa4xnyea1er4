https://aa360cc.cloudatacdn.com/u5kjv4jxytd3sdgge5uogji5dg4huat2pxrm2qibdbm5rmcr6fv2tcniineq/6aqs1w2xla~