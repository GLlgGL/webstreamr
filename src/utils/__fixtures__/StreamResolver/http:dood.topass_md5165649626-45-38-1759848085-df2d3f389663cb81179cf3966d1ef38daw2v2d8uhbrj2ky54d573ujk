https://et1095c.cloudatacdn.com/u5kjxj7s27d3sdgge5woezkbi7milmjrckehcrwuw63gsvytfwxlfmjzvu3a/jli1t787d1~