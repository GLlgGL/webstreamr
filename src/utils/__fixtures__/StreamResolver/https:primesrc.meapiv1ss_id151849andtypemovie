{"servers":[{"name":"Mixdrop","key":"KmF1N","file_size":"2 GB","file_name":"Solaris 1972 PROPER 1080p BluRay x264-SADPANDA - [ENG SUBS] - tt0069293.mp4"},{"name":"Mixdrop","key":"98Gq8","file_size":"2 GB","file_name":"Solaris 1972 PROPER 1080p BluRay x264-SADPANDA - [ENG SUBS] - tt0069293.mp4"},{"name":"Dood","key":"nm5qr","file_size":"1.8 GB","file_name":"Solaris 1972 PROPER 1080p BluRay x264-SADPANDA - ENG SUBS - tt0069293"},{"name":"Dailymotion","key":"-XkLK","file_size":null,"file_name":null},{"name":"YouTube","key":"W9yNA","file_size":null,"file_name":null}],"info":{"tmdb_image":"/uER8n2iitx8Tek6v0wXeF5lMGXp.jpg","tmdb_backdrop":"https://imgproxy.primesrc.me/69ipI6yDMn4pOh35XYMNNSDVdvKcT5DZpCPAYWxAH_o/rs:fit:1000:10000:false/aHR0cHM6Ly9pbWFnZS50bWRiLm9yZy90L3AvdzEyODAvbW0zYU5UakZ2dW8wamVpdG5lUkRjMVpVTmpJLmpwZw","title":"Solaris","status":null,"release_date":"1972-09-26","description":"A psychologist is sent to a station orbiting a distant planet in order to discover what has caused the crew to go insane."}}