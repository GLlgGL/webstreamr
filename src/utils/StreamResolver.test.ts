import { ContentType } from 'stremio-addon-sdk';
import winston from 'winston';
import { BlockedError, HttpError, NotFoundError, QueueIsFullError, TimeoutError, TooManyRequestsError, TooManyTimeoutsError } from '../error';
import { createExtractors, Extractor, ExtractorRegistry } from '../extractor';
import { Source, SourceResult } from '../source';
import { MeineCloud } from '../source/MeineCloud';
import { MostraGuarda } from '../source/MostraGuarda';
import { PrimeWire } from '../source/PrimeWire';
import { createTestContext } from '../test';
import { BlockedReason, CountryCode, Format, UrlResult } from '../types';
import { FetcherMock } from './FetcherMock';
import { ImdbId } from './id';
import { StreamResolver } from './StreamResolver';

const logger = winston.createLogger({ transports: [new winston.transports.Console({ level: 'nope' })] });
const fetcher = new FetcherMock(`${__dirname}/__fixtures__/StreamResolver`);
const ctx = createTestContext({ de: 'on', it: 'on' });

const meineCloud = new MeineCloud(fetcher);
const mostraGuarda = new MostraGuarda(fetcher);
const primeWire = new PrimeWire(fetcher);

describe('resolve', () => {
  test('returns info as stream if no sources were configured', async () => {
    const streamResolver = new StreamResolver(logger, new ExtractorRegistry(logger, createExtractors(fetcher)));

    const streams = await streamResolver.resolve(ctx, [], 'movie', new ImdbId('tt123456789', undefined, undefined));

    expect(streams).toMatchSnapshot();
  });

  test('returns source errors as stream', async () => {
    const fetcherSpy = jest.spyOn(fetcher, 'text').mockRejectedValue('ups, an error occurred.');
    const streamResolver = new StreamResolver(logger, new ExtractorRegistry(logger, createExtractors(fetcher)));

    const streams = await streamResolver.resolve(ctx, [meineCloud], 'movie', new ImdbId('tt123456789', undefined, undefined));
    expect(streams).toMatchSnapshot();

    const streamsWithShowErrors = await streamResolver.resolve({ ...ctx, config: { ...ctx.config, showErrors: 'on' } }, [meineCloud], 'movie', new ImdbId('tt123456789', undefined, undefined));
    expect(streamsWithShowErrors).toMatchSnapshot();

    fetcherSpy.mockRestore();
  });

  test('returns empty array if no source found anything', async () => {
    const streamResolver = new StreamResolver(logger, new ExtractorRegistry(logger, createExtractors(fetcher)));

    const streams = await streamResolver.resolve(ctx, [meineCloud, mostraGuarda], 'movie', new ImdbId('tt12345678', undefined, undefined));

    expect(streams).toMatchSnapshot();
  });

  test('returns empty array if no source supported the type', async () => {
    const streamResolver = new StreamResolver(logger, new ExtractorRegistry(logger, createExtractors(fetcher)));

    const streams = await streamResolver.resolve(ctx, [meineCloud, mostraGuarda], 'series', new ImdbId('tt12345678', 1, 1));

    expect(streams).toMatchSnapshot();
  });

  test('returns sorted results', async () => {
    const streamResolver = new StreamResolver(logger, new ExtractorRegistry(logger, createExtractors(fetcher)));

    const streams = await streamResolver.resolve(ctx, [meineCloud, mostraGuarda], 'movie', new ImdbId('tt29141112', undefined, undefined));
    expect(streams.ttl).not.toBeUndefined();
    expect(streams.streams).toMatchSnapshot();

    const streamsWithExternalUrls = await streamResolver.resolve({ ...ctx, config: { ...ctx.config, includeExternalUrls: 'on' } }, [meineCloud, mostraGuarda], 'movie', new ImdbId('tt29141112', undefined, undefined));
    expect(streamsWithExternalUrls.ttl).not.toBeUndefined();
    expect(streamsWithExternalUrls.streams).toMatchSnapshot();
  });

  test('returns ytId instead of url for YouTube video', async () => {
    const streamResolver = new StreamResolver(logger, new ExtractorRegistry(logger, createExtractors(fetcher)));

    const streams = await streamResolver.resolve({ ...ctx, config: { ...ctx.config, en: 'on' } }, [primeWire], 'movie', new ImdbId('tt0069293', undefined, undefined));
    expect(streams.ttl).not.toBeUndefined();
    expect(streams.streams).toMatchSnapshot();
  });

  test('adds error info', async () => {
    class MockSource extends Source {
      public readonly id = 'mocksource';

      public readonly label = 'MockSource';

      public readonly contentTypes: ContentType[] = ['movie'];

      public readonly countryCodes: CountryCode[] = [CountryCode.de];

      public readonly baseUrl = 'https://example.com';

      public readonly handleInternal = async (): Promise<SourceResult[]> => {
        return [{ url: new URL('https://example.com'), meta: { countryCodes: [CountryCode.de] } }];
      };
    }

    class MockExtractor extends Extractor {
      public readonly id = 'mockextractor';

      public readonly label = 'MockExtractor';

      public override readonly ttl = 1;

      public readonly supports = (): boolean => true;

      protected readonly extractInternal = async (): Promise<UrlResult[]> =>
        [
          {
            url: new URL('https://example.com'),
            format: Format.unknown,
            isExternal: true,
            error: new BlockedError(BlockedReason.cloudflare_challenge, {}),
            label: 'hoster.com',
            sourceId: '',
            meta: {
              countryCodes: [CountryCode.de],
            },
          },
          {
            url: new URL('https://example.com'),
            format: Format.unknown,
            isExternal: true,
            error: new BlockedError(BlockedReason.cloudflare_censor, {}),
            label: 'hoster.com',
            sourceId: '',
            meta: {
              countryCodes: [CountryCode.de],
            },
          },
          {
            url: new URL('https://example.com'),
            format: Format.unknown,
            isExternal: true,
            error: new BlockedError(BlockedReason.media_flow_proxy_auth, {}),
            label: 'hoster.com',
            sourceId: '',
            meta: {
              countryCodes: [CountryCode.de],
            },
          },
          {
            url: new URL('https://example.com'),
            format: Format.unknown,
            isExternal: true,
            error: new BlockedError(BlockedReason.unknown, {}),
            label: 'hoster.com',
            sourceId: '',
            meta: {
              countryCodes: [CountryCode.de],
            },
          },
          {
            url: new URL('https://working2.com'),
            format: Format.unknown,
            label: 'working1',
            sourceId: 'hostercom',
            meta: {
              countryCodes: [CountryCode.de],
            },
          },
          {
            url: new URL('https://example.com'),
            format: Format.unknown,
            isExternal: true,
            error: new TooManyRequestsError(10),
            label: 'hoster.com',
            sourceId: '',
            meta: {
              countryCodes: [CountryCode.de],
            },
          },
          {
            url: new URL('https://example.com'),
            format: Format.unknown,
            isExternal: true,
            error: new TooManyTimeoutsError(),
            label: 'hoster.com',
            sourceId: '',
            meta: {
              countryCodes: [CountryCode.de],
            },
          },
          {
            url: new URL('https://working1.com'),
            format: Format.unknown,
            label: 'working2',
            sourceId: 'hostercom',
            meta: {
              countryCodes: [CountryCode.de],
            },
          },
          {
            url: new URL('https://example2.com'),
            format: Format.unknown,
            isExternal: true,
            error: new TypeError(),
            label: 'hoster.com',
            sourceId: '',
            meta: {
              countryCodes: [CountryCode.de],
            },
          },
          {
            url: new URL('https://example2.com'),
            format: Format.unknown,
            isExternal: true,
            error: new TimeoutError(),
            label: 'hoster.com',
            sourceId: '',
            meta: {
              countryCodes: [CountryCode.de],
            },
          },
          {
            url: new URL('https://example3.com'),
            format: Format.unknown,
            isExternal: true,
            error: new QueueIsFullError(),
            label: 'hoster.com',
            sourceId: '',
            meta: {
              countryCodes: [CountryCode.de],
            },
          },
          {
            url: new URL('https://example4.com'),
            format: Format.unknown,
            isExternal: true,
            error: new HttpError(500, 'Internal Server Error', { 'x-foo': 'bar' }),
            label: 'hoster.com',
            sourceId: '',
            meta: {
              countryCodes: [CountryCode.de],
            },
          },
          {
            url: new URL('https://example5.com'),
            format: Format.unknown,
            isExternal: true,
            error: new HttpError(418, 'I\'m a tea pot', { 'x-foo': 'bar' }),
            label: 'hoster.com',
            sourceId: '',
            meta: {
              countryCodes: [CountryCode.de],
            },
          },
        ];
    }

    const streamResolver = new StreamResolver(logger, new ExtractorRegistry(logger, [new MockExtractor(fetcher)]));

    const streams = await streamResolver.resolve(ctx, [new MockSource()], 'movie', new ImdbId('tt11655566', undefined, undefined));
    expect(streams).toMatchSnapshot();

    const streamsWithShowErrors = await streamResolver.resolve({ ...ctx, config: { ...ctx.config, showErrors: 'on' } }, [new MockSource()], 'movie', new ImdbId('tt11655566', undefined, undefined));
    expect(streamsWithShowErrors).toMatchSnapshot();
  });

  test('ignores not found errors', async () => {
    class MockSource extends Source {
      public readonly id = 'mocksource';

      public readonly label = 'MockSource';

      public readonly contentTypes: ContentType[] = ['movie'];

      public readonly countryCodes: CountryCode[] = [CountryCode.de];

      public readonly baseUrl = 'https://example.com';

      public readonly handleInternal = async (): Promise<SourceResult[]> => {
        throw new NotFoundError();
      };
    }

    const streamResolver = new StreamResolver(logger, new ExtractorRegistry(logger, createExtractors(fetcher)));

    const streams = await streamResolver.resolve(ctx, [new MockSource()], 'movie', new ImdbId('tt12345678', undefined, undefined));

    expect(streams).toMatchSnapshot();
  });
});
