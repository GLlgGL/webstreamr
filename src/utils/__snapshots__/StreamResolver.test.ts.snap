// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`resolve adds error info 1`] = `
{
  "streams": [
    {
      "behaviorHints": {
        "bingeGroup": "webstreamr-hostercom",
        "notWebReady": true,
      },
      "name": "WebStreamr 🇩🇪",
      "title": "🔗 working1",
      "url": "https://working2.com/",
    },
    {
      "behaviorHints": {
        "bingeGroup": "webstreamr-hostercom",
        "notWebReady": true,
      },
      "name": "WebStreamr 🇩🇪",
      "title": "🔗 working2",
      "url": "https://working1.com/",
    },
  ],
}
`;

exports[`resolve adds error info 2`] = `
{
  "streams": [
    {
      "behaviorHints": {
        "notWebReady": true,
      },
      "externalUrl": "https://example5.com/",
      "name": "WebStreamr 🇩🇪",
      "title": "🔗 hoster.com
❌ Request failed with status 418 (I'm a tea pot). Request-id: test.",
    },
    {
      "behaviorHints": {
        "notWebReady": true,
      },
      "externalUrl": "https://example4.com/",
      "name": "WebStreamr 🇩🇪",
      "title": "🔗 hoster.com
❌ Remote server has issues. We can't fix this, please try later again.",
    },
    {
      "behaviorHints": {
        "notWebReady": true,
      },
      "externalUrl": "https://example3.com/",
      "name": "WebStreamr 🇩🇪",
      "title": "🔗 hoster.com
⏳ Request queue is full. Please try again later or consider self-hosting.",
    },
    {
      "behaviorHints": {
        "notWebReady": true,
      },
      "externalUrl": "https://example2.com/",
      "name": "WebStreamr 🇩🇪",
      "title": "🔗 hoster.com
🐢 Request timed out.",
    },
    {
      "behaviorHints": {
        "notWebReady": true,
      },
      "externalUrl": "https://example2.com/",
      "name": "WebStreamr 🇩🇪",
      "title": "🔗 hoster.com
❌ Request failed. Request-id: test.",
    },
    {
      "behaviorHints": {
        "notWebReady": true,
      },
      "externalUrl": "https://example.com/",
      "name": "WebStreamr 🇩🇪",
      "title": "🔗 hoster.com
🚦 Too many recent timeouts. Please try again later.",
    },
    {
      "behaviorHints": {
        "notWebReady": true,
      },
      "externalUrl": "https://example.com/",
      "name": "WebStreamr 🇩🇪",
      "title": "🔗 hoster.com
🚦 Request was rate-limited. Please try again later or consider self-hosting.",
    },
    {
      "behaviorHints": {
        "notWebReady": true,
      },
      "externalUrl": "https://example.com/",
      "name": "WebStreamr 🇩🇪",
      "title": "🔗 hoster.com
⚠️ Request was blocked. Reason: unknown",
    },
    {
      "behaviorHints": {
        "notWebReady": true,
      },
      "externalUrl": "https://example.com/",
      "name": "WebStreamr 🇩🇪",
      "title": "🔗 hoster.com
⚠️ MediaFlow Proxy authentication failed. Please set the correct password.",
    },
    {
      "behaviorHints": {
        "notWebReady": true,
      },
      "externalUrl": "https://example.com/",
      "name": "WebStreamr 🇩🇪",
      "title": "🔗 hoster.com
⚠️ Request was blocked. Reason: cloudflare_censor",
    },
    {
      "behaviorHints": {
        "notWebReady": true,
      },
      "externalUrl": "https://example.com/",
      "name": "WebStreamr 🇩🇪",
      "title": "🔗 hoster.com
⚠️ Request was blocked. Reason: cloudflare_challenge",
    },
    {
      "behaviorHints": {
        "bingeGroup": "webstreamr-hostercom",
        "notWebReady": true,
      },
      "name": "WebStreamr 🇩🇪",
      "title": "🔗 working1",
      "url": "https://working2.com/",
    },
    {
      "behaviorHints": {
        "bingeGroup": "webstreamr-hostercom",
        "notWebReady": true,
      },
      "name": "WebStreamr 🇩🇪",
      "title": "🔗 working2",
      "url": "https://working1.com/",
    },
  ],
}
`;

exports[`resolve ignores not found errors 1`] = `
{
  "streams": [],
  "ttl": 900000,
}
`;

exports[`resolve returns empty array if no source found anything 1`] = `
{
  "streams": [],
  "ttl": 900000,
}
`;

exports[`resolve returns empty array if no source supported the type 1`] = `
{
  "streams": [],
  "ttl": 900000,
}
`;

exports[`resolve returns info as stream if no sources were configured 1`] = `
{
  "streams": [
    {
      "externalUrl": "http://localhost/",
      "name": "WebStreamr",
      "title": "⚠️ No sources found. Please re-configure the plugin.",
    },
  ],
}
`;

exports[`resolve returns sorted results 1`] = `
[
  {
    "behaviorHints": {
      "bingeGroup": "webstreamr-dropload_de",
      "notWebReady": true,
      "videoSize": 1395864371,
    },
    "name": "WebStreamr 🇩🇪 1080p",
    "title": "des-teufels-bad-2024
💾 1.3 GB 🔗 Dropload from MeineCloud",
    "url": "https://srv34.dropload.io/hls2/01/00197/lyo2h1snpe5c_h/master.m3u8?t=MZSqs3F338ocbM2eTs8X23s585dWcYuilGrNNcidNas&s=1749846846&e=14400&f=987607&srv=srv27&i=0.0&sp=0&ii=50.99.15.93&p1=srv27&p2=srv27",
  },
  {
    "behaviorHints": {
      "bingeGroup": "webstreamr-dropload_it",
      "notWebReady": true,
      "videoSize": 1181116006,
    },
    "name": "WebStreamr 🇮🇹 720p",
    "title": "the-devils-bath-2024-sd-sub-ita
💾 1.1 GB 🔗 Dropload from MostraGuarda",
    "url": "https://srv34.dropload.io/hls2/01/00200/xsr90y2ltdyx_h/master.m3u8?t=rsJMeR8rtWZxFi_3DOdlt98U0rvAMincCe4-awDQzHE&s=1749846846&e=14400&f=1004116&srv=srv24&i=0.0&sp=0&ii=50.99.15.93&p1=srv24&p2=srv24",
  },
  {
    "behaviorHints": {
      "bingeGroup": "webstreamr-supervideo_it",
      "notWebReady": true,
      "videoSize": 1181116006,
    },
    "name": "WebStreamr 🇮🇹 720p",
    "title": "the-devils-bath-2024-sd-sub-ita
💾 1.1 GB 🔗 SuperVideo from MostraGuarda",
    "url": "https://hfs311.serversicuro.cc/hls/,dnzpcjj427g4a3gyvbth53rxqgffbu4vol4o7gg7pskpzb6josghbkwhnisq,.urlset/master.m3u8",
  },
  {
    "behaviorHints": {
      "bingeGroup": "webstreamr-supervideo_de",
      "notWebReady": true,
      "videoSize": 1073741824,
    },
    "name": "WebStreamr 🇩🇪 720p",
    "title": "des-teufels-bad-2024
💾 1 GB 🔗 SuperVideo from MeineCloud",
    "url": "https://hfs309.serversicuro.cc/hls/,dnzpfi3d27g4a3gyvbmh5klwtl65qh654hyjtgupd6p577fze64mz3bsmdyq,.urlset/master.m3u8",
  },
  {
    "behaviorHints": {
      "bingeGroup": "webstreamr-doodstream_de",
      "notWebReady": true,
      "proxyHeaders": {
        "request": {
          "Referer": "http://dood.to",
        },
      },
      "videoSize": 1395864371,
    },
    "name": "WebStreamr 🇩🇪",
    "title": "des-teufels-bad-2024
💾 1.3 GB 🔗 DoodStream from MeineCloud",
    "url": "https://aa360cc.cloudatacdn.com/u5kjv4jxytd3sdgge5uogji5dg4huat2pxrm2qibdbm5rmcr6fv2tcniineq/6aqs1w2xla~mocked-random-string?token=7uebebipnnhusa4xnyea1er4&expiry=639837296000",
  },
  {
    "behaviorHints": {
      "bingeGroup": "webstreamr-doodstream_it",
      "notWebReady": true,
      "proxyHeaders": {
        "request": {
          "Referer": "http://dood.to",
        },
      },
      "videoSize": 1181116006,
    },
    "name": "WebStreamr 🇮🇹",
    "title": "the-devils-bath-2024-sd-sub-ita
💾 1.1 GB 🔗 DoodStream from MostraGuarda",
    "url": "https://et1095c.cloudatacdn.com/u5kjxj7s27d3sdgge5woezkbi7milmjrckehcrwuw63gsvytfwxlfmjzvu3a/jli1t787d1~mocked-random-string?token=aw2v2d8uhbrj2ky54d573ujk&expiry=639837296000",
  },
]
`;

exports[`resolve returns sorted results 2`] = `
[
  {
    "behaviorHints": {
      "bingeGroup": "webstreamr-dropload_de",
      "notWebReady": true,
      "videoSize": 1395864371,
    },
    "name": "WebStreamr 🇩🇪 1080p",
    "title": "des-teufels-bad-2024
💾 1.3 GB 🔗 Dropload from MeineCloud",
    "url": "https://srv34.dropload.io/hls2/01/00197/lyo2h1snpe5c_h/master.m3u8?t=MZSqs3F338ocbM2eTs8X23s585dWcYuilGrNNcidNas&s=1749846846&e=14400&f=987607&srv=srv27&i=0.0&sp=0&ii=50.99.15.93&p1=srv27&p2=srv27",
  },
  {
    "behaviorHints": {
      "bingeGroup": "webstreamr-dropload_it",
      "notWebReady": true,
      "videoSize": 1181116006,
    },
    "name": "WebStreamr 🇮🇹 720p",
    "title": "the-devils-bath-2024-sd-sub-ita
💾 1.1 GB 🔗 Dropload from MostraGuarda",
    "url": "https://srv34.dropload.io/hls2/01/00200/xsr90y2ltdyx_h/master.m3u8?t=rsJMeR8rtWZxFi_3DOdlt98U0rvAMincCe4-awDQzHE&s=1749846846&e=14400&f=1004116&srv=srv24&i=0.0&sp=0&ii=50.99.15.93&p1=srv24&p2=srv24",
  },
  {
    "behaviorHints": {
      "bingeGroup": "webstreamr-supervideo_it",
      "notWebReady": true,
      "videoSize": 1181116006,
    },
    "name": "WebStreamr 🇮🇹 720p",
    "title": "the-devils-bath-2024-sd-sub-ita
💾 1.1 GB 🔗 SuperVideo from MostraGuarda",
    "url": "https://hfs311.serversicuro.cc/hls/,dnzpcjj427g4a3gyvbth53rxqgffbu4vol4o7gg7pskpzb6josghbkwhnisq,.urlset/master.m3u8",
  },
  {
    "behaviorHints": {
      "bingeGroup": "webstreamr-supervideo_de",
      "notWebReady": true,
      "videoSize": 1073741824,
    },
    "name": "WebStreamr 🇩🇪 720p",
    "title": "des-teufels-bad-2024
💾 1 GB 🔗 SuperVideo from MeineCloud",
    "url": "https://hfs309.serversicuro.cc/hls/,dnzpfi3d27g4a3gyvbmh5klwtl65qh654hyjtgupd6p577fze64mz3bsmdyq,.urlset/master.m3u8",
  },
  {
    "behaviorHints": {
      "bingeGroup": "webstreamr-doodstream_de",
      "notWebReady": true,
      "proxyHeaders": {
        "request": {
          "Referer": "http://dood.to",
        },
      },
      "videoSize": 1395864371,
    },
    "name": "WebStreamr 🇩🇪",
    "title": "des-teufels-bad-2024
💾 1.3 GB 🔗 DoodStream from MeineCloud",
    "url": "https://aa360cc.cloudatacdn.com/u5kjv4jxytd3sdgge5uogji5dg4huat2pxrm2qibdbm5rmcr6fv2tcniineq/6aqs1w2xla~mocked-random-string?token=7uebebipnnhusa4xnyea1er4&expiry=639837296000",
  },
  {
    "behaviorHints": {
      "bingeGroup": "webstreamr-doodstream_it",
      "notWebReady": true,
      "proxyHeaders": {
        "request": {
          "Referer": "http://dood.to",
        },
      },
      "videoSize": 1181116006,
    },
    "name": "WebStreamr 🇮🇹",
    "title": "the-devils-bath-2024-sd-sub-ita
💾 1.1 GB 🔗 DoodStream from MostraGuarda",
    "url": "https://et1095c.cloudatacdn.com/u5kjxj7s27d3sdgge5woezkbi7milmjrckehcrwuw63gsvytfwxlfmjzvu3a/jli1t787d1~mocked-random-string?token=aw2v2d8uhbrj2ky54d573ujk&expiry=639837296000",
  },
  {
    "behaviorHints": {
      "bingeGroup": "webstreamr-external_de",
      "notWebReady": true,
    },
    "externalUrl": "https://mixdrop.ag/e/3nzwveprim63or6",
    "name": "WebStreamr 🇩🇪 ⚠️ external",
    "title": "🔗 mixdrop.ag from MeineCloud",
  },
  {
    "behaviorHints": {
      "bingeGroup": "webstreamr-external_it",
      "notWebReady": true,
    },
    "externalUrl": "https://mixdrop.ag/e/vk196d6xfzwwo1",
    "name": "WebStreamr 🇮🇹 ⚠️ external",
    "title": "🔗 mixdrop.ag from MostraGuarda",
  },
]
`;

exports[`resolve returns source errors as stream 1`] = `
{
  "streams": [],
}
`;

exports[`resolve returns source errors as stream 2`] = `
{
  "streams": [
    {
      "externalUrl": "https://meinecloud.click",
      "name": "WebStreamr",
      "title": "🔗 MeineCloud
❌ Request failed. Request-id: test.",
    },
  ],
}
`;

exports[`resolve returns ytId instead of url for YouTube video 1`] = `
[
  {
    "behaviorHints": {
      "bingeGroup": "webstreamr-doodstream_en",
      "notWebReady": true,
      "proxyHeaders": {
        "request": {
          "Referer": "http://dood.to",
        },
      },
      "videoSize": 1932735283,
    },
    "name": "WebStreamr 🇺🇸",
    "title": "Solaris 1972 PROPER 1080p BluRay x264-SADPANDA - ENG SUBS - tt0069293
💾 1.8 GB 🔗 DoodStream from PrimeWire",
    "url": "https://do365ki.cloudatacdn.com/u5kjzcovmpd3sdgge7tc4jacdzaikwer6ff4qd463b6qj6lodlgpnsbkffvq/xy59ygfkku~mocked-random-string?token=aej8q7ajug1gt0nd4lbs8zgd&expiry=639837296000",
  },
  {
    "behaviorHints": {
      "bingeGroup": "webstreamr-youtube_en",
      "notWebReady": true,
    },
    "name": "WebStreamr 🇺🇸",
    "title": "Solaris | SCIENCE FICTION | FULL MOVIE | directed by Tarkovsky
🔗 YouTube from PrimeWire",
    "ytId": "Z8ZhQPaw4rE",
  },
]
`;
